File: csharp/components/StrataAdministration/Diagnostic/AbstractDiagnosticResult.cs
//  ##########################################################################
//  # File Name: AbstractDiagnosticResult.cs
//  # Copyright: 2013, Sapientia Systems, LLC.
//  ##########################################################################
 
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;

namespace Strata.Administration.Diagnostic
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// Base class of all <c>IDiagnosticResult</c> types.
    /// </summary>
    /// 
    /// <author>JFL</author>
    /// <conventions>$conventionspath$</conventions>
    ///  
    public abstract 
    class AbstractDiagnosticResult:
        IDiagnosticResult
    {
	    private readonly IList<IDiagnosticReporter> reporters;
	    private readonly IList<DiagnosticOutput>    contents;
	    private readonly DiagnosticOutputGenerator  generator;
	
        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// Creates a new <c>AbstractDiagnosticResult</c> instance.
        /// </summary>
        /// 
	    protected 
	    AbstractDiagnosticResult()
	    {
		    reporters = new List<IDiagnosticReporter>();
		    contents  = new List<DiagnosticOutput>();
		    generator = new DiagnosticOutputGenerator();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see 
        /// cref="IDiagnosticResult.AttachReporter(IDiagnosticReporter)"/>
        /// </summary>
        /// 
	    public void 
	    AttachReporter(IDiagnosticReporter reporter)
	    {
		    if ( !HasReporter( reporter ) )
			    reporters.Add( reporter );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see 
        /// cref="IDiagnosticResult.DetachReporter(IDiagnosticReporter)"/>
        /// </summary>
        /// 
	    public void 
	    DetachReporter(IDiagnosticReporter reporter)
	    {
		    reporters.Remove( reporter );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.GetReporters()"/>
        /// </summary>
        /// 
	    public IList<IDiagnosticReporter> 
	    GetReporters()
	    {
		    return new ReadOnlyCollection<IDiagnosticReporter>( reporters );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.HasReporters()"/>
        /// </summary>
        /// 
	    public bool
	    HasReporters()
	    {
		    return reporters.Count != 0;
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.HasReporter(IDiagnosticReporter)"/>
        /// </summary>
        /// 
	    public bool 
	    HasReporter(IDiagnosticReporter reporter)
	    {
		    return reporters.Contains( reporter );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.BeginReports()"/>
        /// </summary>
        /// 
	    public virtual void 
	    BeginReports()
	    {
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.BeginReport();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.EndReports()"/>
        /// </summary>
        /// 
	    public virtual void 
	    EndReports()
	    {
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.EndReport();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver.BeginDiagnostic(IDiagnostic)"/>
        /// </summary>
        /// 
	    public virtual void 
	    BeginDiagnostic(IDiagnostic d)
	    {
		    DiagnosticOutput output = null;
		
		    generator.DiagnosticName = d.Name;
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.BeginDiagnostic( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver.EndDiagnostic(IDiagnostic)"/>
        /// </summary>
        /// 
	    public virtual void 
	    EndDiagnostic(IDiagnostic d)
	    {
		    DiagnosticOutput output =  generator.GenerateOutput();
		
		    contents.Add( output );
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.EndDiagnostic( output );
		
		    generator.Clear();
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver
        ///                .ReportBeginFailure(
        ///                    IDiagnostic,DiagnosticAbortedException)"/>
        /// </summary>
        /// 
	    public virtual void 
	    ReportBeginFailure(IDiagnostic d,DiagnosticAbortedException e)
	    {
		    DiagnosticOutput output = null;
		
		    generator.ResultState = DiagnosticResultState.ABORTED;
		    generator.Description = "diagnostic failed to run.";
		    generator.Exception = e ;
		
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.ReportBeginFailure( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver
        ///                .ReportCheckSuccess(IDiagnostic,String)"/>
        /// </summary>
        /// 
	    public virtual void 
	    ReportCheckSuccess(IDiagnostic d,String msg)
	    {
		    DiagnosticOutput output = null;
		
		    generator.ResultState = DiagnosticResultState.SUCCEEDED;
		    generator.Description = msg;
		    generator.Exception = null;
		
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.ReportCheckSuccess( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver
        ///                .ReportCheckFailure(
        ///                    IDiagnostic,DiagnosticException)"/>
        /// </summary>
        /// 
	    public virtual void 
	    ReportCheckFailure(IDiagnostic d,DiagnosticException e)
	    {
		    DiagnosticOutput output = null;
		
		    generator.ResultState = DiagnosticResultState.FAILED;
		    generator.Description = "diagnostic failed with exception:";
		    generator.Exception   = e ;
		
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.ReportCheckFailure( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver
        ///                .ReportRecoverySuccess(
        ///                    IDiagnostic,String)"/>
        /// </summary>
        /// 
	    public virtual void 
	    ReportRecoverySuccess(IDiagnostic d,String msg)
	    {
		    DiagnosticOutput output = null;
		
		    generator.ResultState = DiagnosticResultState.RECOVERED;
		    generator.Description = msg;
		    generator.Exception   = null;
		
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.ReportRecoverySuccess( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver
        ///                .ReportRecoveryFailure(
        ///                    IDiagnostic,DiagnosticException)"/>
        /// </summary>
        /// 
	    public virtual void 
	    ReportRecoveryFailure(IDiagnostic d,DiagnosticException e)
	    {
		    DiagnosticOutput output = null;

		    generator.ResultState = DiagnosticResultState.FAILED;
		    generator.Description =  
		        "diagnostic recovery failed with exception:";
		    generator.Exception   = e;
		
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.ReportRecoveryFailure( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticReceiver
        ///                .ReportUnknownFailure(IDiagnostic d,Exception)"/>
        /// </summary>
        /// 
	    public virtual void 
	    ReportUnknownFailure(IDiagnostic d,Exception e)
	    {
		    DiagnosticOutput output = null;
		
		    generator.ResultState = DiagnosticResultState.FAILED;
		    generator.Description = "diagnostic failed with exception:";
		    generator.Exception   = e;
		
		    output = generator.GenerateOutput();
		
		    foreach (IDiagnosticReporter reporter in reporters)
			    reporter.ReportUnknownFailure( output );
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.GetNumberOfDiagnostics()"/>
        /// </summary>
        /// 
	    public int 
	    GetNumberOfDiagnostics()
	    {
		    return contents.Count;
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.GetNumberOfAborts()"/>
        /// </summary>
        /// 
	    public int 
	    GetNumberOfAborts()
	    {
            return 
                contents
                    .Count(
                        o => o.ResultState == DiagnosticResultState.ABORTED);
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.GetNumberOfSuccesses()"/>
        /// </summary>
        /// 
	    public int 
	    GetNumberOfSuccesses()
	    {
            return 
                contents
                    .Count(
                        o => o.ResultState == DiagnosticResultState.SUCCEEDED);
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.GetNumberOfFailures()"/>
        /// </summary>
        /// 
	    public int 
	    GetNumberOfFailures()
	    {
            return 
                contents
                    .Count(
                        o => o.ResultState == DiagnosticResultState.FAILED);
	    }

        //////////////////////////////////////////////////////////////////////
        /// <summary>
        /// <see cref="IDiagnosticResult.GetContents()"/>
        /// </summary>
        /// 
	    public IList<DiagnosticOutput> 
	    GetContents()
	    {
		    return new ReadOnlyCollection<DiagnosticOutput>(contents);
	    }

    }
}

//  ##########################################################################

File: java/Components/Strata.Diagnostic.Core/src/strata/diagnostic/core/common/AbstractDiagnosticResult.java
// ##########################################################################
// # File Name:	AbstractDiagnosticResult.java
// #
// # Copyright:	2011, Sapientia Systems, LLC. All Rights Reserved.
// #
// # License:	This file is part of the StrataCommon Framework.
// #
// #   			The StrataInspector Framework is free software: you 
// #			can redistribute it and/or modify it under the terms of 
// #			the GNU Lesser General Public License as published by
// #    		the Free Software Foundation, either version 3 of the 
// #			License, or (at your option) any later version.
// #
// #    		The StrataInspector Framework is distributed in the 
// #			hope that it will be useful, but WITHOUT ANY WARRANTY; 
// #			without even the implied warranty of MERCHANTABILITY or 
// #			FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser 
// #			General Public License for more details.
// #
// #    		You should have received a copy of the GNU Lesser 
// #			General Public License along with the StrataInspector
// #			Framework. If not, see http://www.gnu.org/licenses/.
// ##########################################################################

package strata.diagnostic.core.common;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;

/**
 * 
 * @author 		
 *     Sapientia Systems 
 * @conventions	
 *     <a href="{@docRoot}/NamingConventions.html">Naming Conventions</a>
 */
public abstract 
class AbstractDiagnosticResult 
	implements IDiagnosticResult
{
	private List<IDiagnosticReporter>  itsReporters;
	private List<DiagnosticOutput>    itsContents;
	private DiagnosticOutputGenerator itsGenerator;
	
	/************************************************************************
	 * Creates a new AbstractDiagnosticResult. 
	 *
	 */
	public 
	AbstractDiagnosticResult()
	{
		itsReporters = new LinkedList<IDiagnosticReporter>();
		itsContents  = new LinkedList<DiagnosticOutput>();
		itsGenerator = new DiagnosticOutputGenerator();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#attachReporter(IDiagnosticReporter)
	 */
	@Override
	public void 
	attachReporter(IDiagnosticReporter reporter)
	{
		if ( !hasReporter( reporter ) )
			itsReporters.add( reporter );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#detachReporter(IDiagnosticReporter)
	 */
	@Override
	public void 
	detachReporter(IDiagnosticReporter reporter)
	{
		itsReporters.remove( reporter );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#getReporters()
	 */
	@Override
	public List<IDiagnosticReporter> 
	getReporters()
	{
		return new LinkedList<IDiagnosticReporter>( itsReporters );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#hasReporters()
	 */
	@Override
	public boolean 
	hasReporters()
	{
		return !itsReporters.isEmpty();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#hasReporter(IDiagnosticReporter)
	 */
	@Override
	public boolean 
	hasReporter(IDiagnosticReporter reporter)
	{
		return itsReporters.contains( reporter );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#beginReports()
	 */
	@Override
	public void 
	beginReports()
	{
		for (IDiagnosticReporter reporter:getReporters())
			reporter.beginReport();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#endReports()
	 */
	@Override
	public void 
	endReports()
	{
		for (IDiagnosticReporter reporter:getReporters())
			reporter.endReport();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#beginDiagnostic(IDiagnostic)
	 */
	@Override
	public void 
	beginDiagnostic(IDiagnostic d)
	{
		DiagnosticOutput output = null;
		
		itsGenerator.setDiagnosticName( d.getName() );
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.beginDiagnostic( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#endDiagnostic(IDiagnostic)
	 */
	@Override
	public void 
	endDiagnostic(IDiagnostic d)
	{
		DiagnosticOutput output =  itsGenerator.generateOutput();
		
		itsContents.add( output );
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.endDiagnostic( output );
		
		itsGenerator.clear();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#reportBeginFailure(
	 * 			IDiagnostic,
	 * 			DiagnosticAbortedException)
	 */
	@Override
	public void 
	reportBeginFailure(IDiagnostic d,DiagnosticAbortedException e)
	{
		DiagnosticOutput output = null;
		
		itsGenerator.setResultState( DiagnosticResultState.ABORTED );
		itsGenerator.setDescription( "IDiagnostic failed to run." );
		itsGenerator.setException( e );
		
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.reportBeginFailure( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#reportCheckSuccess(IDiagnostic,String)
	 */
	@Override
	public void 
	reportCheckSuccess(IDiagnostic d,String msg)
	{
		DiagnosticOutput output = null;
		
		itsGenerator.setResultState( DiagnosticResultState.SUCCEEDED );
		itsGenerator.setDescription( msg );
		itsGenerator.setException( null );
		
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.reportCheckSuccess( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#reportCheckFailure(
	 * 			IDiagnostic,
	 * 			DiagnosticException)
	 */
	@Override
	public void 
	reportCheckFailure(IDiagnostic d,DiagnosticException e)
	{
		DiagnosticOutput output = null;
		
		itsGenerator.setResultState( DiagnosticResultState.FAILED );
		itsGenerator.setDescription( "IDiagnostic failed with exception:" );
		itsGenerator.setException( e );
		
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.reportCheckFailure( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#reportRecoverySuccess(IDiagnostic,String)
	 */
	@Override
	public void 
	reportRecoverySuccess(IDiagnostic d,String msg)
	{
		DiagnosticOutput output = null;
		
		itsGenerator.setResultState( DiagnosticResultState.RECOVERED );
		itsGenerator.setDescription( msg );
		itsGenerator.setException( null );
		
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.reportRecoverySuccess( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#reportRecoveryFailure(
	 *			IDiagnostic,
	 *			DiagnosticException)
	 */
	@Override
	public void 
	reportRecoveryFailure(IDiagnostic d,DiagnosticException e)
	{
		DiagnosticOutput output = null;

		itsGenerator.setResultState( DiagnosticResultState.FAILED );
		itsGenerator.setDescription( 
		    "IDiagnostic recovery failed with exception:" );
		itsGenerator.setException( e );
		
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.reportRecoveryFailure( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticReceiver#reportUnknownFailure(IDiagnostic,Exception)
	 */
	@Override
	public void 
	reportUnknownFailure(IDiagnostic d,Exception e)
	{
		DiagnosticOutput output = null;
		
		itsGenerator.setResultState( DiagnosticResultState.FAILED );
		itsGenerator.setDescription( "IDiagnostic failed with exception:" );
		itsGenerator.setException( e );
		
		output = itsGenerator.generateOutput();
		
		for (IDiagnosticReporter reporter:getReporters())
			reporter.reportUnknownFailure( output );
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#getNumberOfDiagnostics()
	 */
	@Override
	public int 
	getNumberOfDiagnostics()
	{
		return itsContents.size();
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#getNumberOfAborts()
	 */
	@Override
	public int 
	getNumberOfAborts()
	{
		int aborts = 0;
		
		for (DiagnosticOutput output:getContents())
			if ( output.getResultState() == DiagnosticResultState.ABORTED )
				++aborts;
		
		return aborts;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#getNumberOfSuccesses()
	 */
	@Override
	public int 
	getNumberOfSuccesses()
	{
		int successes = 0;
		
		for (DiagnosticOutput output:getContents())
			if ( output.getResultState() == DiagnosticResultState.SUCCEEDED )
				++successes;
		
		return successes;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#getNumberOfFailures()
	 */
	@Override
	public int 
	getNumberOfFailures()
	{
		int failures = 0;
		
		for (DiagnosticOutput output:getContents())
			if ( output.getResultState() == DiagnosticResultState.FAILED )
				++failures;
		
		return failures;
	}

	/************************************************************************
	 * {@inheritDoc} 
	 * @see IDiagnosticResult#getContents()
	 */
	@Override
	public List<DiagnosticOutput> 
	getContents()
	{
		return Collections.unmodifiableList( itsContents );
	}

	
}


// ##########################################################################
